components/tidb_query/src/rpn_expr/impl_json.rs

```diff
@@ -14,6 +16,83 @@ fn json_type(arg: &Option<Json>) -> Result<Option<Bytes>> {
         .map(|json_arg| Bytes::from(json_arg.json_type())))
 }
 
+#[rpn_fn(varg)]
+#[inline]
+fn json_array(args: &[&Option<Json>]) -> Result<Option<Json>> {
+    Ok(Some(Json::Array(
+        args.iter()
+            .map(|json| match json {
+                None => Json::None,
+                Some(json) => json.to_owned(),
+            })
+            .collect(),
+    )))
+}
+
+fn json_validator(expr: &tipb::Expr) -> Result<()> {
+    for chunk in expr.get_children().chunks(2) {
+        if chunk.len() == 1 {
+            return Err(other_err!(
+                "Incorrect parameter count in the call to native function 'JSON_OBJECT'"
+            ));
+        } else {
+            super::function::validate_expr_return_type(&chunk[0], Bytes::EVAL_TYPE)?;
+            super::function::validate_expr_return_type(&chunk[1], Json::EVAL_TYPE)?;
+        }
+    }
+    Ok(())
+}
+
+/// Required args like `&[(&Option<Byte>, &Option<Json>)]`.
+#[rpn_fn(raw_varg, extra_validator = json_validator)]
+#[inline]
+fn json_object(raw_args: &[ScalarValueRef]) -> Result<Option<Json>> {
+    let mut pairs = BTreeMap::new();
+    for chunk in raw_args.chunks(2) {
+        // chunk.len() must be 1 or 2 here.
+        let key: &Option<Bytes> = Evaluable::borrow_scalar_value_ref(&chunk[0]);
+        let key = match key {
+            // json_object should raise an error if key is None(NULL)
+            None => Err(other_err!(
+                "Data truncation: JSON documents may not contain NULL member names."
+            )),
+            Some(v) => {
+                String::from_utf8(v.to_owned()).map_err(|e| crate::codec::Error::from(e).into())
+            }
+        }?;
+
+        let value: &Option<Json> = Evaluable::borrow_scalar_value_ref(&chunk[1]);
+        let value = match value {
+            None => Json::None,
+            Some(v) => v.to_owned(),
+        };
+        pairs.insert(key, value);
+    }
+    Ok(Some(Json::Object(pairs)))
+}
+
+macro_rules! parse_opt {
+    ($expr:expr) => {{
+        match $expr {
+            Some(ref v) => v.to_owned().clone(),
```

Comment:
Why we need `.clone()`?

Reply:
> Why we need `.clone()`?

`to_owned` is already `A generalization of Clone to borrowed data.` So you're right, I removed `.clone` now.